docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,313 @@
+# Meaning of `?` variables in K
+
+## Summary
+
+The frontend syntax uses question mark variables of the form `?X` to refer to
+existential variables, and uses `ensures` to specify logical constraints on
+those variables.
+These variables are only allowed to appear in the RHS of a rule.
+
+When K rules are translated to Kore axioms, the `?` variables are existentially
+quantified at the top of the RHS and the `ensures` clause is and-ed to the RHS.
+In addition, the regular non-`?` variables are universally quantified at the top
+of the entire rule.
+When both `?`-variables and regular variables are present, we quantify regular
+variables first, followed by `?`-variables.
+
+This design document presents four typical language constructs or mathematical
+functions that have unspecified or nondeterministic behaviors and shows their
+axiomatization in Kore.
+
+We also propose the corresponding K frontend syntax to define some of these
+behaviors.
+
+### Implementation concerns
+
+Except for the `fresh` variants (Example B), none of the other are currently
+supported by either the K frontend of the Haskell backend.
+
+## Example A: Random Number Construct `rand()`
+
+The random number construct `rand()` is a language construct which could be
+easily conceived to be part of the syntax of a programming language:
+
+```
+Exp ::= "rand" "(" ")"
+```
+
+### Axioms
+
+The intended semantics of `rand()` is that it can rewrite to any integer in
+a single step.
+Here is its axiom
+
+```
+(RAND) \forall I:Int . rand() -> * I
+// equivalent form(s)
+//     rand() -> \forall I:Int . * I
+//     rand() -> * I
+// non-equivalent form(s)
+//     rand() -> * (\forall I:Int . I)
+```
+
+Note that the `(RAND)` axiom is logically equivalent to the following infinitely
+many axioms.
+
+```
+(RAND0)  rand() -> * 0
+(RAND1)  rand() -> * 1
+(RAND2)  rand() -> * 2
+  ...    ...
+(RAND-1) rand() -> * (-1)
+(RAND-2) rand() -> * (-2)
+  ...    ...
+```
+
+One can define variants of `rand()` by further constraining the output variable
+as a precondition to the rule.
+
+#### Rand-like examples
+
+1. `randBounded(M,N)` can rewrite to any integer between `M` and `N`
+    ```
+    \forall M:Int.\forall N:Int.\forall I:Int.
+        I >= M /\ I <= N -> randBounded(M, N) -> * I
+    ```
+
+1. `randInList(Is)` takes a list `Is` of items
+   and can rewrite in one step to any item in `Is`.
+
+    ```
+    \forall Is:List . \forall I . I \in Is -> randInList(Is) -> * I
```

Comment:
We probably want a different symbol here, because `\in` is the Kore alias for unification:

```suggestion
    \forall Is:List . \forall I . inList(I, Is) -> randInList(Is) -> * I
```